### PR TITLE
Implement tab completion

### LINKED ISSUE #27: Tab completion support
## Description

I use tab completion quite a lot in GHCi. Alas, it doesn’t seem to work in GHCiTUI: pressing the Tab key shows `:complete <my input>` at the bottom of the Source pane, but it doesn’t actually show the completion in the GHCi pane. I’m not sure if this is due to a bug, or simply because it hasn’t been implemented yet — in which case, I’d like to request it if possible!

### To Reproduce

1. Open `ghcitui` and switch to the GHCi pane.
2. Type input and press the Tab key.

### Expected behaviour

The input is completed as in GHCi.

### Screenshots

Example of what happens after I press Tab:

image

### Environment

- OS: Linux: EndeavourOS (essentially Arch Linux)
- GHCiTUI Version: 0.1.0.0

### PR DESCRIPTION
**Pull request by bradrn**
Closes #27.

This is just a first go at it — I’m quite sure my code can be improved. In particular, I’m still a little confused about how GHCiTUI manages state. That being said, when I test this code, it does seem to work.

### CODE REVIEW

**Comment by CrystalSplitter** [comment]
Thank you so much for writing this! I'm going to check it out. Might do a bit of rewriting, but I really do appreciate anyone taking an interest in the project! At a first glance this looks great.

**Comment by CrystalSplitter** [comment]
As for how it handles state:

You may be getting confused between the Brick state (which is accessed while in the `EventM` monad), versus the `AppState` which contains everything not handled by Brick (and a few things in Brick we need to manually handle too...). I don't really use AppState in a monadic way mostly because when I tried it, I felt code readability was harmed with very little gain.

Basically you get the current AppState with `B.get`, do operations on it (pure or impure), then you put it back in with `B.put`. Then handle the Brick state itself (such as caching, visual updates, etc.)

**Comment by bradrn** [comment]
> You may be getting confused between the Brick state (which is accessed while in the `EventM` monad), versus the `AppState` which contains everything not handled by Brick (and a few things in Brick we need to manually handle too...).

No, it was more about the meaning of the various components of `AppState`, `AppInterpState`, `InterpState`, etc., and precisely how they’re used. For instance, I can’t figure out the point of `cmdHistory`.

**Comment by CrystalSplitter** [comment]
Oh, I see. I'll explain from the inside-out then (you don't need to update this PR, but this will help me write documentation). The gist is anything prefixed with "App*" is about the TUI.

* The `Daemon` or the `InterpState` is the connection point between GHCiTUI and GHCiD. It handles all GHCi interactions that do not require the TUI. It holds:
  * The GHCi Handle
  * The execution context.
  * The current paused location.
  * The state of the program trace.
  * The current bindings.
  * The modules to files mapping.
  * A number of other things that I don't think are actually used anymore...

  It does not understand anything about the TUI, Events, Widgets, etc. It does not depend on `Brick` at all, and is intended to work equally as well with a TUI as well as a full GUI.

* The `AppInterpState` represents the live interpreter panel (the one at the bottom). It's solely visual TUI state. It holds bits of Brick state, and only Brick related things. like the Brick.editor, the last things you ran in the terminal, whether you're scrolling the history. At one point, `AppInterpState` was just record entries in `AppState`, but this got messy to manage.
  To sum, `AppInterpState` is just the state of the live interpreter panel and nothing else.

* The `AppState` is the state of the entire application. It holds the `InterpState`, the `AppInterpState`, and every other bit of stateful information that's required to run the TUI that's not covered by Brick. It contains:
  * What's displayed on the source window.
  * What's the start up config.
  * What's the active window.
  * What's the scale of the windows.
  * What's in the debug logs.
  * What's the current working directory.
  
Does that help explain it? If so, I'll add these to the docs :)

**Comment by CrystalSplitter** [comment]
Alright, played around with this a fair bit. You've done the core work, though I'm thinking of making a number of changes. These are:

1. I'm going to move any handling of `:complete` and parsing its out put to `ghcitui-core` as I don't think it's dependent on `Brick`. Things which are front-end agnostic should go there.
2. I'm going to present completions as a separate widget below the input line. This is because I want the option to display/format autocompletion recommendations differently. Specifically line wrap and/or width aware tabular format. Plus it doesn't pollute the history this way.
3. I'm going to replace all the partial functions that I can't prove will not hit their partiality condition.
4. I'm thinking of having a rotary state, much like `fish` shell's autocomplete function. Not sure if I'll do this one, 

These three are going to take a little bit of time, so feel free to continue to use your own fork while I see what I can do.

**Comment by bradrn** [comment]
> * I'm going to move any handling of `:complete` and parsing its out put to `ghcitui-core` as I don't think it's dependent on `Brick`. Things which are front-end agnostic should go there.

Sounds reasonable!

> * I'm going to present completions as a separate widget below the input line. This is because I want the option to display/format autocompletion recommendations differently. Specifically line wrap and/or width aware tabular format. Plus it doesn't pollute the history this way.

I actually did want to do this, but I’m unfamiliar with Brick, and the architecture confused me. (On which note, your previous post helped a bit, thanks!)

> * I'm going to replace all the partial functions that I can't prove will not hit their partiality condition.

At least in the parsing code, this currently works on the assumption that GHCi `:complete repl` output will always adhere to the same format. Whether this is a bad assumption or not, I don’t know.

> * I'm thinking of having a rotary state, much like `fish` shell's autocomplete function. Not sure if I'll do this one

I’m not familiar with `fish`, how does that work?

**Comment by CrystalSplitter** [comment]
> I’m not familiar with fish, how does that work?

Imagine every time you press tab consecutively, there's a stateful "tab counter" that increments. Pressing it once displays the recommendations. Pressing it twice auto-selects the first recommendation. Pressing it twice auto-selects the second recommendation, and so on. Holding tab cycles through all the recommendations.

**Comment by CrystalSplitter** [comment]
Incorporated this PR into #32! Thank you again!

**Comment by bradrn** [comment]
> 2\. I'm going to present completions as a separate widget below the input line. This is because I want the option to display/format autocompletion recommendations differently. Specifically line wrap and/or width aware tabular format. Plus it doesn't pollute the history this way.

Trying out your PR, it looks like you didn’t implement it… which is fine, it’s not at all critical, but we might want to keep it in mind as something nice to possibly implement in the future.

**Comment by CrystalSplitter** [comment]
Yeah, I think it would have been a neat idea, but it turned out a bit harder than I anticipated. In particular, handling a large number of options gets difficult and it needs a custom scrolling functionality (a hidden viewport may work with visibility requests?). For example, if the completions were too long, it would scroll the GHCi viewport, hiding what you were typing. Bits of this work ended up in the final result, but not anything stateful.

After trying it for a few days, I decided to give up the idea for now. The ergonomics just weren't feeling right, and it's very important to me that any feature is comfortable to use, works nearly all the time and can't ever make the experience worse.

It would be nice, I agree. It would be even nicer if it was just part of GHCi by default though. :)
